porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go

```diff
@@ -90,349 +90,159 @@ func (r *PackageVariantSetReconciler) Reconcile(ctx context.Context, req ctrl.Re
-		if err != nil {
-			return nil, err
-		}
-		if repo == "" {
-			return nil, fmt.Errorf("error evaluating repo name: received empty string")
+			objSel = &api.ObjectSelector{
+				LabelSelector: *target.RepositorySelector,
+				APIVersion:    configapi.TypeRepository.APIVersion(),
+				Kind:          configapi.TypeRepository.Kind,
+			}
 		}
-		result = append(result, &pkgvarapi.Downstream{
-			Package: downstreamPackageName,
-			Repo:    repo,
-		})
-	}
-	return result, nil
-}
-
-func (r *PackageVariantSetReconciler) getSelectedObjects(ctx context.Context, selectors []api.Selector) (map[resid.ResId]*yaml.RNode, error) {
-	selectedObjects := make(map[resid.ResId]*yaml.RNode) // this is a map to prevent duplicates
-
-	for _, selector := range selectors {
+		// a selector against a set of arbitrary objects
 		uList := &unstructured.UnstructuredList{}
-		group, version := resid.ParseGroupVersion(selector.APIVersion)
+		group, version := resid.ParseGroupVersion(objSel.APIVersion)
 		uList.SetGroupVersionKind(schema.GroupVersionKind{
 			Group:   group,
 			Version: version,
-			Kind:    selector.Kind,
+			Kind:    objSel.Kind,
 		})
 
-		opts := []client.ListOption{client.InNamespace(selector.Namespace)}
-		if selector.Labels != nil {
-			labelSelector, err := metav1.LabelSelectorAsSelector(selector.Labels)
-			if err != nil {
-				return nil, err
-			}
-			opts = append(opts, client.MatchingLabelsSelector{Selector: labelSelector})
+		opts := []client.ListOption{client.InNamespace(pvs.Namespace)}
+		labelSelector, err := metav1.LabelSelectorAsSelector(&objSel.LabelSelector)
+		if err != nil {
+			return nil, err
 		}
+		opts = append(opts, client.MatchingLabelsSelector{Selector: labelSelector})
+
 		if err := r.Client.List(ctx, uList, opts...); err != nil {
 			return nil, err
 		}
 
-		for _, u := range uList.Items {
-			objAsRNode, err := r.convertObjectToRNode(&u)
-			if err != nil {
-				return nil, fmt.Errorf("error converting unstructured object to RNode: %v", err)
-			}
-			if fnruntime.IsMatch(objAsRNode, selector.ToKptfileSelector()) {
-				selectedObjects[resid.FromRNode(objAsRNode)] = objAsRNode
-			}
+		// TODO: fire event; set condition?
+		if len(uList.Items) == 0 {
+			klog.Warningf("no objects selected by spec.targets[%d]", i)
 		}
-	}
-	return selectedObjects, nil
-}
-
-func (r *PackageVariantSetReconciler) getDownstreamPackageName(targetName *api.PackageName,
-	upstreamPackageName string,
-	obj *yaml.RNode) (string, error) {
-
-	if targetName == nil {
-		return upstreamPackageName, nil
-	}
-
-	packageName, err := r.fetchValue(targetName.Name, obj)
-	if err != nil {
-		return "", err
-	}
-	if packageName == "" {
-		packageName = upstreamPackageName
-	}
-
-	suffix, err := r.fetchValue(targetName.NameSuffix, obj)
-	if err != nil {
-		return "", err
-	}
+		for _, u := range uList.Items {
+			result = append(result, pvContext{
+				template:       target.Template,
+				repoDefault:    u.GetName(),
```

Comment:
If selectors matches arbitrary objects, is it still correct to use the name of the resource here?

Reply:
Yes; at least, that is how it is specified in the design :).

This is just the default. The user can override it using an expression based on the object (for example, `target.annotations['repo-name']`). That comes later in the PackageVariantTemplate.